ld.lld 20 on win32: error: unknown argument: --push-state

Hello,

easily reproducible in a MSYS2 shell or with the llvm-mingw toolchain:

```
$ ld.lld -m i386pep --subsystem windows --shared -Bdynamic -o score.dll foo.o --push-state --pop-state
lld: error: unknown argument: --push-state
lld: error: unknown argument: --pop-state
```

It worked fine with previous versions. push-state / pop-state was implemented here: https://github.com/llvm/llvm-project/commit/f75ea0b99580f3ffc7473f6703890181ad231b83 and `ld.lld --help` mentions the options.

CMake runs linker detection code in <https://github.com/Kitware/CMake/blob/v4.3.2/Modules/Platform/Linker/GNU.cmake>. That code directly invokes the underlying linker program.

Got a test case for repro.

CMakeLists.txt:
```cmake
cmake_minimum_required(VERSION 4.2)
project(test C CXX)
add_library(a a.cxx)
add_executable(b a.cxx)
target_link_libraries(b PRIVATE $<LINK_LIBRARY:WHOLE_ARCHIVE,a>)
```

tc.cmake
```cmake
set(CMAKE_SYSTEM_NAME Windows)
set(CMAKE_SYSTEM_PROCESSOR i686)
set(CMAKE_C_COMPILER clang)
set(CMAKE_CXX_COMPILER clang++)
set(CMAKE_C_COMPILER_TARGET i686-pc-windows-gnu)
set(CMAKE_CXX_COMPILER_TARGET i686-pc-windows-gnu)
set(CMAKE_PROGRAM_PATH ${CMAKE_CURRENT_LIST_DIR}/tc/bin)
```

a.cxx is empty.

Toolchain files are in tc.

```
cmake -S . -B build -G Ninja --toolchain=tc.cmake
cmake --build build
```

Output:
```
[4/4] Linking CXX executable b.exe
FAILED: [code=1] b.exe 
C:\WINDOWS\system32\cmd.exe /C "cd . && D:\cmake_llvm_repro\tc\bin\clang++.exe --target=i686-pc-windows-gnu   CMakeFiles/b.dir/a.cxx.obj -o b.exe -Wl,--out-implib,libb.dll.a -Wl,--major-image-version,0,--minor-image-version,0  -Xlinker --push-state -Xlinker --whole-archive  liba.a  -Xlinker --pop-state  -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -luuid -lcomdlg32 -ladvapi32 && cd ."
lld: error: unknown argument: --push-state
lld: error: unknown argument: --pop-state
clang-22: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
```

--- a/lld/MinGW/Driver.cpp
+++ b/lld/MinGW/Driver.cpp
@@ -44,6 +44,7 @@
 #include "llvm/TargetParser/Host.h"
 #include "llvm/TargetParser/Triple.h"
 #include <optional>
+#include <stack>
 
 using namespace lld;
 using namespace llvm::opt;
@@ -576,6 +577,11 @@ bool link(ArrayRef<const char *> argsArr, llvm::raw_ostream &stdoutOS,
 
   StringRef prefix = "";
   bool isStatic = false;
+  struct PushPopState {
+    StringRef prefix;
+    bool isStatic;
+  };
+  std::stack<PushPopState, std::vector<PushPopState>> pushPopStates;
   for (auto *a : args) {
     switch (a->getOption().getID()) {
     case OPT_INPUT:
@@ -603,6 +609,18 @@ bool link(ArrayRef<const char *> argsArr, llvm::raw_ostream &stdoutOS,
       break;
     case OPT_Bdynamic:
       isStatic = false;
+      break;
+    case OPT_push_state:
+      pushPopStates.push({prefix, isStatic});
+      break;
+    case OPT_pop_state:
+      if (pushPopStates.empty()) {
+        error("unbalanced --push-state/--pop-state");
+        break;
+      }
+      prefix = pushPopStates.top().prefix;
+      isStatic = pushPopStates.top().isStatic;
+      pushPopStates.pop();
       break;
     }
   }

--- a/lld/MinGW/Options.td
+++ b/lld/MinGW/Options.td
@@ -125,6 +125,10 @@ def o: JoinedOrSeparate<["-"], "o">, MetaVarName<"<path>">,
   HelpText<"Path to file to write output">;
 defm out_implib: Eq<"out-implib", "Import library name">;
 defm output_def: Eq<"output-def", "Output def file">;
+def pop_state: F<"pop-state">,
+  HelpText<"Restore the states saved by --push-state">;
+def push_state: F<"push-state">,
+  HelpText<"Save the current state of --Bstatic/--Bdynamic and --whole-archive">;
 defm reloc_section: B_enable_disable<"reloc-section",
      "Enable base relocations", "Disable base relocations">;
 defm section_alignment: Eq<"section-alignment", "Set section alignment">;

--- a/lld/docs/ReleaseNotes.rst
+++ b/lld/docs/ReleaseNotes.rst
@@ -49,6 +49,10 @@ COFF Improvements
 MinGW Improvements
 ------------------
 
+* Added ``--push-state`` and ``--pop-state``, offering the same semantics as
+  when used with the ELF linker: The state of ``--Bstatic``/``--Bdynamic`` and
+  ``--whole-archive`` are pushed onto a stack and popped from it.
+
 MachO Improvements
 ------------------
 

--- a/lld/test/MinGW/driver.test
+++ b/lld/test/MinGW/driver.test
@@ -501,3 +501,11 @@ NO-FAT-LTO: -fat-lto-objects:no
 
 RUN: ld.lld -### -m i386pep foo.o --no-fat-lto-objects --fat-lto-objects 2>&1 | FileCheck -check-prefix FAT-LTO %s
 FAT-LTO: -fat-lto-objects{{ }}
+
+RUN: ld.lld -### -m i386pep foo.o --push-state --whole-archive bar.a --pop-state baz.a 2>&1 | FileCheck -check-prefix PUSH-POP-STATE %s
+RUN: ld.lld -### -m i386pep foo.o -push-state -whole-archive bar.a -pop-state baz.a 2>&1 | FileCheck -check-prefix PUSH-POP-STATE %s
+PUSH-POP-STATE: foo.o -wholearchive:bar.a baz.a
+
+RUN: not ld.lld -m i386pep foo.o --pop-state 2>&1 | FileCheck -check-prefix PUSH-POP-STATE-UNBALANCED %s
+RUN: not ld.lld -m i386pep foo.o --push-state --pop-state --pop-state 2>&1 | FileCheck -check-prefix PUSH-POP-STATE-UNBALANCED %s
+PUSH-POP-STATE-UNBALANCED: error: unbalanced --push-state/--pop-state

